miden-lib/asm/sat/internal/faucet.masm

```diff
@@ -0,0 +1,281 @@
+#! against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a fungible faucet.
+#! - If the fungible asset being minted is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the asset is not well formed.
+#! - If the total issuance after minting is greater than the maximum amount allowed.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - amount is the amount of the fungible asset to mint.
+#! - ASSET is the asset that was minted.
+export.mint_fungible_asset
+    # assert that we are executing a transaction against a fungible faucet (access checks)
+    exec.account::get_id dup exec.account::is_fungible_faucet assert
+    # => [fungible_faucet_id, ASSET]
+
+    # assert that the asset is associated with the faucet the transaction is being executed against
+    dup.1 assert_eq
+    # => [ASSET]
+
+    # assert the asset is well formed
+    dup.1 not assert 
+    dup.2 not assert 
+    # => [ASSET]
+
+    # get the current total issuance
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::get_item 
+    # => [TOTAL_ISSUANCE, ASSET]
+
+    # prepare stack to ensure that minting the asset will not exceed the maximum
+    dup.7 dup exec.asset::get_fungible_asset_max_amount dup.3
+    # => [total_issuance, max_allowed_issuance, amount, amount, TOTAL_ISSUANCE, ASSET]
+
+    # compute difference to ensure that the total issuance will not exceed the maximum
+    sub lte assert
+    # => [amount, TOTAL_ISSUANCE, ASSET]
+
+    # update the total issuance
+    add push.FAUCET_DATA_STORAGE_SLOT exec.account::set_item dropw
+    # => [ASSET]
+
+    # add the asset to the input vault for asset preservation checks
+    dupw exec.layout::get_input_vault_root_ptr movdn.4 exec.asset_vault::add_fungible_asset dropw
+    # => [ASSET]
+end
+
+#! Burns a fungible asset associated with the fungible faucet the transaction is being executed
+#! against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a fungible faucet.
+#! - If the fungible asset being burned is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the asset is not well formed.
+#! - If the amount being burned is greater than the total input to the transaction.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - ASSET is the asset that was burned.
+proc.burn_fungible_asset
+    # assert that we are executing a transaction against a fungible faucet (access checks)
+    exec.account::get_id dup exec.account::is_fungible_faucet assert
+    # => [fungible_faucet_id, ASSET]
+
+    # assert that the asset is associated with the faucet the transaction is being executed against
+    dup.1 assert_eq
+    # => [ASSET]
+
+    # assert the asset is well formed
+    dup.1 not assert
+    dup.2 not assert 
+    # => [ASSET]
+
+    # fetch TOTAL_ISSUANCE such that we can compute the new total issuance
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::get_item
+    # => [TOTAL_ISSUANCE, ASSET]
```

Comment:
Below on line 96 we call `remove_fungible_asset` on the `input_vault`.  This will check to assert that the we can only burn an amount up to the amount of the asset provided as input to the transaction.  This is a stricter condition than comparing against the total issuance. Still happy to introduce the check against the total issuance if you think it's best?

Reply:
I was thinking that if there is a bug somewhere, a note could be created such that it would carry a greater balance than total issuance. If such a note was a part of input notes, the check against input vault would still pass. Granted, this would be a pretty catastrophic bug in the first place, but I still think might be good to add an extra check since it is not too expensive (and also since burning assets will probably be much more rare than minting them).

Reply:
Makes sense I will add it.